### ISSUE TITLE
Prune data blob missing

### ISSUE THREAD

**Issue by tgelter** [restic/restic#3241]
Output of `restic version`
--------------------------
`restic 0.11.0 compiled with go1.15.3 on linux/amd64`

How did you run restic exactly?
-------------------------------
```[root@guapuranas bin]# ./call_restic bowsnties.env prune
Sourcing /root/bin/bowsnties.env
repository 2afda69a opened successfully, password is correct
counting files in repo
building new index for repo
[3:53:35] 100.00%  413683 / 413683 packs
repository contains 413683 packs (1608699 blobs) with 1.986 TiB
processed 1608699 blobs: 524 duplicate blobs, 748.142 MiB duplicate
load all snapshots
find data that is still in use for 40 snapshots
[1:22] 100.00%  40 / 40 snapshots
Fatal: [<data/95688dee> <data/a819f792> <data/9a124f36> <data/c75f3034> <data/9902f413> <data/9e87b47e>] not found in the new index
Data blobs seem to be missing, aborting prune to prevent further data loss!
Please report this error (along with the output of the 'prune' run) at
https://github.com/restic/restic/issues/new/choose
```
```[root@guapuranas bin]# cat call_restic 
#!/bin/bash

ENV_FILES_PATH=/root/bin

if [[ $# -eq 0 ]]; then
    	echo "Usage: $0 <env_vars> <action>"
	exit 1
fi

echo "Sourcing ${ENV_FILES_PATH}/${1}"
# shellcheck source=/dev/null
source $ENV_FILES_PATH/$1

if [[ $2 =~ [Bb]ackup ]]; then
	if [[ $MAX_UPLOAD_RATE -eq 0 ]]; then
		restic backup --exclude-caches -x "$PATH_TO_BACKUP"
	else
		restic --limit-upload "$MAX_UPLOAD_RATE" backup --exclude-caches -x "$PATH_TO_BACKUP"
	fi
fi

if [[ $2 =~ [Cc]lean ]]; then
	restic forget \
	${RESTIC_KEEP_HOURLY:+--keep-hourly ${RESTIC_KEEP_HOURLY}} \
	${RESTIC_KEEP_DAILY:+--keep-daily ${RESTIC_KEEP_DAILY}} \
	${RESTIC_KEEP_WEEKLY:+--keep-weekly ${RESTIC_KEEP_WEEKLY}} \
	${RESTIC_KEEP_MONTHLY:+--keep-monthly ${RESTIC_KEEP_MONTHLY}} \
	${RESTIC_KEEP_YEARLY:+--keep-yearly ${RESTIC_KEEP_YEARLY}}
fi

if [[ $2 =~ [Cc]heck ]]; then
	restic check
fi

if [[ $2 =~ [Pp]rune ]]; then
	restic prune
fi

if [[ $2 =~ [Ss]napshots ]]; then
	restic snapshots
fi

if [[ $2 =~ [Ss]tats ]]; then
	restic stats
fi

if [[ $2 =~ [Uu]nlock ]]; then
	restic unlock
fi```
```[root@guapuranas bin]# cat bowsnties.env 
export RESTIC_REPOSITORY="s3:http://redacted:9000/tgelter-bowsnties" # Example: s3:http://<ipaddr>:9000/restic-agelter-test
export RESTIC_PASSWORD="redacted"          # Example: myPassword
export PATH_TO_BACKUP="/mnt/nas_zfs_vol0/bowsnties/"                       # Example: /backup/file
export AWS_SECRET_ACCESS_KEY="redacted"    # Example foo/bar
export AWS_ACCESS_KEY_ID="redacted"                            # Example: blehKeyId 
#export MAX_UPLOAD_RATE="500"         					   # Example: 100 (Kbps)
export MAX_UPLOAD_RATE=""         					   # Example: 100 (Kbps)
export RESTIC_KEEP_HOURLY="1"                                              # Number of hourly snapshots to keep (keep empty to ignore)
export RESTIC_KEEP_DAILY="30"                                              # Number of daily snapshots to keep (keep empty to ignore)
export RESTIC_KEEP_WEEKLY="4"                                              # Number of weekly snapshots to keep (keep empty to ignore)
export RESTIC_KEEP_MONTHLY="12"                                            # Number of monthly snapshots to keep (keep empty to ignore)
export RESTIC_KEEP_YEARLY=                                                 # Number of yearly snapshots to keep (keep empty to ignore)
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Remote Minio server (more details upon request)

Expected behavior
-----------------
The prune should succeed.

Actual behavior
---------------
The prune failed and recommended that I submit a bug

Steps to reproduce the behavior
-------------------------------
Attempt to prune my own personal large (1.986 TiB) repository.

Do you have any idea what may have caused this?
-----------------------------------------------
Unknown, but possibly data loss on the remote end which has recently encountered a power outage / unclean server shut down.

Do you have an idea how to solve the issue?
-------------------------------------------
No

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Restic has served me well for a couple years now, this is the first issue I've encountered aside from the occasional repo lock issue where I have to manually unlock the repository to unblock further snapshotting.

**Comment by tgelter** [comment]
I ran `restic check` this morning and was presented with:
```
error for tree 806ae803:
  tree 806ae803: file "IMG_6735.CR2" blob 11 size could not be found
  tree 806ae803: file "IMG_6736.CR2" blob 3 size could not be found
  tree 806ae803: file "IMG_6736.CR2" blob 15 size could not be found
  tree 806ae803: file "IMG_6737.CR2" blob 13 size could not be found
  tree 806ae803: file "IMG_6738.CR2" blob 4 size could not be found
  tree 806ae803: file "IMG_6739.CR2" blob 1 size could not be found
  tree 806ae803, blob 95688dee: not found in index
  tree 806ae803, blob c75f3034: not found in index
  tree 806ae803, blob 9e87b47e: not found in index
  tree 806ae803, blob 9902f413: not found in index
  tree 806ae803, blob a819f792: not found in index
  tree 806ae803, blob 9a124f36: not found in index
Fatal: repository contains errors
```
I'm running `restic rebuild-index` now and will report back my findings.

**Comment by tgelter** [comment]
Results of the rebuild-index:
```
[root@guapuranas bin]# ./call_restic bowsnties.env rebuild-index
Sourcing /root/bin/bowsnties.env
repository 2afda69a opened successfully, password is correct
counting files in repo
[2:11:25] 100.00%  415135 / 415135 packs
finding old index files
saved new indexes as [115be3e5 bac8d414 0a157714 e1918c68 20631754 62b54f7b 51479af5 aadf146c 0f4cda5f c7902f14 d5138523 04c4b9c3 bc5c9d56 d6914788 001b6251 83c022ca 8a197557 4bd9aac3 2f3ebe67 add539d1 7330927a ee77921c e0689c40 26ea6344 2e3edb98 fc6f5fe9 88b3ca91 4393579e b1bb6993 7e0e734a 395d5e50 ca23dfab 89781ff2 47a19b8a 53fd27db b6ba274b 2a61ef1a 0a1df970 82f692ac d47f2b1e f5174341 a454653d 55faa566 7fa147e7 f873950c f0b5dd65 822ef560 bfe0c208 89397461 c8e75a5d 09d784ab f9d3f837 546430fc eed1610d 66719162 4e0792ab 91dda021 fad7cbbb 64f6890f b105d335 22893a91 ea6c12c0 ae2deba5 23f9c05c e017a4f0 6855df3a a33f85f1 659da0f5 2862618c 07b5eaec 992aa3f9 1808e637 a1d91870 24d82820 35772c74 b156c0cc ce1b9b40 fb7a1e49 98233b35 40d1b3ad 007cf582 5612eb7c b190dbcc f2cbb4c2 93e2695e 0c1675ec 214e011b 4addaf28 5cea361e 9db5e596 080dc110 f35e726e e2cf022c de35d624 81286957 62b62402 ed898f86 6bfb1b63 3648869b 784cdc20 05b8fca8 d2bd1180 9d05ff97 cc83eee6 a328c5e3 2ad6a44b 1bb0c170 10af53bb b1a3f458 cf162384 0d45dda5 363efd96 4da2eac4 3d15ac14 7b82c84a 83279392 12778dba 24a1b5b4 921eaedf b64c688f 564ced43 feb9b47d ec3a1f13 d8681c43 52dd3bca bce687f6 36459f47 aaf94386 41e127c4 54ff00a3 9aad090b 66c63ed3 903be983 a6ac9a74 39db1199 7407cced 262597a0 8f0edfb7 549eab6b]
remove 1367 old index files
[0:16] 100.00%  1367 / 1367 files deleted
```
Results of the following backup job:
```
Jan 28 13:00:27 guapuranas.gelter.com systemd[1]: Starting backup of bowsnties...
Jan 28 13:00:27 guapuranas.gelter.com call_restic[735024]: Sourcing /root/bin/bowsnties.env
Jan 28 15:07:09 guapuranas.gelter.com call_restic[735026]: Files:         125 new,     0 changed, 158741 unmodified
Jan 28 15:07:09 guapuranas.gelter.com call_restic[735026]: Dirs:            2 new,     3 changed,  6622 unmodified
Jan 28 15:07:09 guapuranas.gelter.com call_restic[735026]: Added to the repo: 2.501 GiB
Jan 28 15:07:09 guapuranas.gelter.com call_restic[735026]: processed 158866 files, 1.243 TiB in 2:06:40
Jan 28 15:07:09 guapuranas.gelter.com call_restic[735026]: snapshot 2fbb68a1 saved
Jan 28 15:07:09 guapuranas.gelter.com systemd[1]: run-backup@bowsnties.service: Succeeded.
Jan 28 15:07:09 guapuranas.gelter.com systemd[1]: Finished backup of bowsnties.
```
Next I'll run check/prune again and see how things look...

**Comment by tgelter** [comment]
Not looking good...
```[root@guapuranas bin]# ./call_restic bowsnties.env check
Sourcing /root/bin/bowsnties.env
using temporary cache in /tmp/restic-check-cache-511777093
repository 2afda69a opened successfully, password is correct
created new cache in /tmp/restic-check-cache-511777093
create exclusive lock for repository
load indexes
check all packs
check snapshots, trees and blobs
error for tree 806ae803:
  tree 806ae803: file "IMG_6735.CR2" blob 11 size could not be found
  tree 806ae803: file "IMG_6736.CR2" blob 3 size could not be found
  tree 806ae803: file "IMG_6736.CR2" blob 15 size could not be found
  tree 806ae803: file "IMG_6737.CR2" blob 13 size could not be found
  tree 806ae803: file "IMG_6738.CR2" blob 4 size could not be found
  tree 806ae803: file "IMG_6739.CR2" blob 1 size could not be found
  tree 806ae803, blob 95688dee: not found in index
  tree 806ae803, blob c75f3034: not found in index
  tree 806ae803, blob 9e87b47e: not found in index
  tree 806ae803, blob 9902f413: not found in index
  tree 806ae803, blob a819f792: not found in index
  tree 806ae803, blob 9a124f36: not found in index
Fatal: repository contains errors
```

**Comment by tgelter** [comment]
After reading about the errors listed in my previous comment [here](https://forum.restic.net/t/unable-to-complete-check-fatal-repository-contains-errors/1244/10), I'm proceeding with another `restic prune` attempt.

**Comment by MichaelEischer** [comment]
@tgelter As `check` still reports errors, `prune` won't work. Please run `restic find --tree 806ae803` to find the affected snapshots. As you've mentioned a server power outage, was the snapshot created a short time before/after that outage? If you still have the log files: Did any of these backup run report errors?

Do you still have the missing file?

**Comment by tgelter** [comment]
You were right about the `prune` not working:
```[root@guapuranas bin]# ./call_restic bowsnties.env check
Sourcing /root/bin/bowsnties.env
using temporary cache in /tmp/restic-check-cache-511777093
repository 2afda69a opened successfully, password is correct
created new cache in /tmp/restic-check-cache-511777093
create exclusive lock for repository
load indexes
check all packs
check snapshots, trees and blobs
error for tree 806ae803:
  tree 806ae803: file "IMG_6735.CR2" blob 11 size could not be found
  tree 806ae803: file "IMG_6736.CR2" blob 3 size could not be found
  tree 806ae803: file "IMG_6736.CR2" blob 15 size could not be found
  tree 806ae803: file "IMG_6737.CR2" blob 13 size could not be found
  tree 806ae803: file "IMG_6738.CR2" blob 4 size could not be found
  tree 806ae803: file "IMG_6739.CR2" blob 1 size could not be found
  tree 806ae803, blob 95688dee: not found in index
  tree 806ae803, blob c75f3034: not found in index
  tree 806ae803, blob 9e87b47e: not found in index
  tree 806ae803, blob 9902f413: not found in index
  tree 806ae803, blob a819f792: not found in index
  tree 806ae803, blob 9a124f36: not found in index
Fatal: repository contains errors
[root@guapuranas bin]# ./call_restic bowsnties.env prune
Sourcing /root/bin/bowsnties.env
repository 2afda69a opened successfully, password is correct
counting files in repo
building new index for repo
[1:51:36] 100.00%  415644 / 415644 packs
repository contains 415644 packs (1615622 blobs) with 1.996 TiB
processed 1615622 blobs: 524 duplicate blobs, 748.142 MiB duplicate
load all snapshots
find data that is still in use for 40 snapshots
[1:25] 100.00%  40 / 40 snapshots
Fatal: [<data/a819f792> <data/c75f3034> <data/9902f413> <data/95688dee> <data/9a124f36> <data/9e87b47e>] not found in the new index
Data blobs seem to be missing, aborting prune to prevent further data loss!
Please report this error (along with the output of the 'prune' run) at
https://github.com/restic/restic/issues/new/choose
```
Here are all of the logs I have for backup runs, I'm not sure if there's anything in particular that would be helpful for me to `grep` for to save you some time looking through them:
[bowsnties-backup-logs.zip](https://github.com/restic/restic/files/5896345/bowsnties-backup-logs.zip)

I have the `restic find --tree 806ae803` command running now. I'll report back on that when it's finished.

**Comment by tgelter** [comment]
```[root@guapuranas bin]# restic find --tree 806ae803
repository 2afda69a opened successfully, password is correct
Found tree 806ae803d072ff801f195ee711f214324e6a82ff12faaab7d5babe870788261a
 ... path /mnt/nas_zfs_vol0/bowsnties/redacted names wedding 2017
 ... in snapshot cf272036 (2019-12-31 23:00:15)
```

**Comment by MichaelEischer** [comment]
Hmm, as the snapshot is rather old, that makes it unlikely that it has been broken since end of 2019. It would be helpful to get an idea of when the repository was damaged. How often do you run prune / check? When did you upgrade to restic 0.11.0? Was the failing prune run the first one since upgrading? I've noticed a few retried file uploads in the log file but nothing too unusual.

If you still have the file with the missing blobs then running `restic backup <filename>` should add the missing file parts back to the repository. If the tree is only contained in that one snapshot, then you could also just remove that snapshot.

**Comment by tgelter** [comment]
Hello again @MichaelEischer, thanks for the response.

I don't have any scheduled jobs for prune/check, I run the following timers for periodic back-up and clean-up (`restic forget`) tasks:
```
[root@guapuranas ~]# systemctl list-timers | grep backup
Mon 2021-02-01 16:00:00 MST 55min left          Mon 2021-02-01 15:00:27 MST 3min 48s ago hourly-backup-clean@bowsnties.timer  run-backup-clean@bowsnties.service
Mon 2021-02-01 16:00:00 MST 55min left          Mon 2021-02-01 15:00:27 MST 3min 48s ago hourly-backup-clean@family.timer     run-backup-clean@family.service   
Mon 2021-02-01 16:00:00 MST 55min left          Mon 2021-02-01 15:00:27 MST 3min 48s ago hourly-backup@bowsnties.timer        run-backup@bowsnties.service      
Mon 2021-02-01 16:00:00 MST 55min left          Mon 2021-02-01 15:00:27 MST 3min 48s ago hourly-backup@family.timer           run-backup@family.service         
Mon 2021-02-08 00:00:00 MST 6 days left         Mon 2021-02-01 00:00:27 MST 15h ago      weekly-backup-check@bowsnties.timer  run-backup-check@bowsnties.service
Mon 2021-02-08 00:00:00 MST 6 days left         Mon 2021-02-01 00:00:27 MST 15h ago      weekly-backup-check@family.timer     run-backup-check@family.service   
Mon 2021-03-01 02:00:00 MST 3 weeks 6 days left Mon 2021-02-01 02:00:27 MST 13h ago      monthly-backup-prune@bowsnties.timer run-backup-prune@bowsnties.service
Mon 2021-03-01 02:00:00 MST 3 weeks 6 days left Mon 2021-02-01 02:00:27 MST 13h ago      monthly-backup-prune@family.timer    run-backup-prune@family.service  
```

Here is the `restic` versioning history from this server:
```
[root@guapuranas ~]# grep restic /var/log/pacman.log 
[2018-01-02 14:33] [PACMAN] Running 'pacman --color auto -U /tmp/yaourt-tmp-tgelter/PKGDEST.vDZ/restic-0.8.1-1-x86_64.pkg.tar.xz'
[2018-01-02 14:33] [ALPM] installed restic (0.8.1-1)
[2018-02-21 22:16] [PACMAN] Running 'pacman --color auto -U /tmp/yaourt-tmp-tgelter/PKGDEST.pM0/restic-0.8.2-1-x86_64.pkg.tar.xz'
[2018-02-21 22:16] [ALPM] upgraded restic (0.8.1-1 -> 0.8.2-1)
[2018-03-04 23:01] [PACMAN] Running 'pacman --color auto -U /tmp/yaourt-tmp-tgelter/PKGDEST.8XP/restic-0.8.3-1-x86_64.pkg.tar.xz'
[2018-03-04 23:01] [ALPM] upgraded restic (0.8.2-1 -> 0.8.3-1)
[2018-04-03 16:10] [ALPM] upgraded restic (0.8.3-1 -> 0.8.3-2)
[2018-04-12 20:48] [ALPM] upgraded restic (0.8.3-2 -> 0.8.3-3)
[2018-05-31 08:19] [ALPM] upgraded restic (0.8.3-3 -> 0.9.0-1)
[2018-06-13 11:40] [ALPM] upgraded restic (0.9.0-1 -> 0.9.1-1)
[2018-08-13 19:19] [ALPM] upgraded restic (0.9.1-1 -> 0.9.2-1)
[2018-11-02 19:41] [ALPM] upgraded restic (0.9.2-1 -> 0.9.3-1)
[2019-01-07 08:53] [ALPM] upgraded restic (0.9.3-1 -> 0.9.4-1)
[2019-02-14 09:40] [ALPM] upgraded restic (0.9.4-1 -> 0.9.4-2)
[2019-05-01 20:50] [ALPM] upgraded restic (0.9.4-2 -> 0.9.5-1)
[2019-08-18 18:54] [ALPM] upgraded restic (0.9.5-1 -> 0.9.5-2)
[2019-10-14 20:22] [ALPM] upgraded restic (0.9.5-2 -> 0.9.5-3)
[2019-11-25T10:13:27-0700] [ALPM] upgraded restic (0.9.5-3 -> 0.9.6-1)
[2020-02-29T15:31:12-0700] [ALPM] upgraded restic (0.9.6-1 -> 0.9.6-2)
[2020-06-10T09:01:19-0600] [ALPM] upgraded restic (0.9.6-2 -> 0.9.6-3)
[2020-09-21T11:53:59-0600] [ALPM] upgraded restic (0.9.6-3 -> 0.10.0-1)
[2020-11-06T08:56:13-0700] [ALPM] upgraded restic (0.10.0-1 -> 0.11.0-1)
```

Just prior to submitting this issue was the first time I had tried to run `restic prune` or `restic check`, so I'm not sure I can provide more information about when the issue would have first been detectable.

I went ahead and ran `restic backup` after copying the directory to the name identified by `restic find` and am now re-running `restic check` and possibly `restic prune` afterward. I'll keep you posted, thanks!

**Comment by tgelter** [comment]
The `check` worked OK, I'm running a `prune` now.
```
[root@guapuranas bin]# ./call_restic bowsnties.env check
Sourcing /root/bin/bowsnties.env
using temporary cache in /tmp/restic-check-cache-343590027
repository 2afda69a opened successfully, password is correct
created new cache in /tmp/restic-check-cache-343590027
create exclusive lock for repository
load indexes
check all packs
check snapshots, trees and blobs
no errors were found
```

**Comment by tgelter** [comment]
The `prune` went well, thanks very much for the help @MichaelEischer!
```
[root@guapuranas bin]# ./call_restic bowsnties.env prune
Sourcing /root/bin/bowsnties.env
repository 2afda69a opened successfully, password is correct
counting files in repo
building new index for repo
[2:43:23] 100.00%  416344 / 416344 packs
repository contains 416344 packs (1618084 blobs) with 1.999 TiB
processed 1618084 blobs: 524 duplicate blobs, 748.142 MiB duplicate
load all snapshots
find data that is still in use for 40 snapshots
[1:15] 100.00%  40 / 40 snapshots
found 987659 of 1618084 data blobs still in use, removing 630425 blobs
will remove 0 invalid files
will delete 169905 packs and rewrite 1346 packs, this frees 842.860 GiB
[3:02:13] 100.00%  1346 / 1346 packs rewritten
counting files in repo
[2:51:16] 100.00%  245702 / 245702 packs
finding old index files
saved new indexes as [0d2cbeeb 3bfa24a9 980fe646 8b1b0a11 ce562dc5 647f08c2 406169de 8062ff4d bfd23479 df9529e8 39191422 491e98c2 85f1a5f0 f550bb10 72ee2ec5 8622b589 9ab8d24c 6fea91e9 93b3695b 71a74fd5 b2119edf f40e1ee2 c6bc0b24 4725843e 6d235fc4 229096c2 887953aa 9f54d9b6 6892dbf9 1a43d759 990216f6 3d56ebab c83deedd 0b9c2105 2953a6b6 04efbd5a 097b5466 11ad1de2 4c8c80a4 d4bd56b4 6256518e d5b18995 93e3f66f b5addfc8 584ac1ae c81ad3bc 3b5af4fe 9ea920db c82ff24f fa22efb2 4fa44f77 745e1768 a6230146 45178a48 710436a7 411da49e 2a81a669 ba61f1da 524372aa 953fc890 ae19df6d 0156a0af 9f1aafcb dee169e5 e180442f 1bf9d953 b49baf6a 0c3f4ed8 19ca85fb 8b5dbde3 b20fd419 42506e10 7ba85fdf 572aa1fe a54e5a41 b9b2ff6a 8cd6fcc8 0e3c7ab0 0fc39fa3 22ddfa53 63e3b440 823dcc32]
remove 168 old index files
[0:02] 100.00%  168 / 168 files deleted
remove 171251 old packs
[40:41] 100.00%  171251 / 171251 files deleted
done
```
Post-`prune` sanity `check`:
```
[root@guapuranas bin]# ./call_restic bowsnties.env check
Sourcing /root/bin/bowsnties.env
using temporary cache in /tmp/restic-check-cache-035420989
repository 2afda69a opened successfully, password is correct
created new cache in /tmp/restic-check-cache-035420989
create exclusive lock for repository
load indexes
check all packs
check snapshots, trees and blobs
no errors were found
```

**Comment by MichaelEischer** [comment]
> Just prior to submitting this issue was the first time I had tried to run restic prune or restic check, so I'm not sure I can provide more information about when the issue would have first been detectable.

Hmm, that makes it indeed rather hard to make a good guess when and how the damage could have occurred. So thanks anyways for the feedback. If the problem shows up again, feel free to open a new issue.
